### ISSUE TITLE
Microsoft Graph support

### ISSUE THREAD

**Issue by hermanius** [swoosh/swoosh#744]
Are there (short-term) plans to support a Microsoft Graph adapter?
https://learn.microsoft.com/en-us/graph/api/user-sendmail?view=graph-rest-1.0&tabs=http

**Comment by princemaple** [comment]
No plans, happy to accept PR 😊

**Comment by geeksilva97** [comment]
Can I take a look?

**Comment by princemaple** [comment]
😃 sure

**Comment by justindotpub** [comment]
I have a working implementation for this and was pondering if a PR would be accepted.

Would you accept a PR that only supports auth using oauth 2.0 tokens, and that shields swoosh from that logic by having the config be `access_token_fn`, an anonymous function that returns the access token to be used in the authorization header?  This allows the user of this adapter to handle oauth 2.0 access token requests in whatever manner they see fit, including dealing with caching tokens and refreshing before their expiration.

**Comment by justindotpub** [comment]
🤔 I'm pondering how to incorporate that into the code in a generic way.  As long as authorization details are always ultimately incorporated into the `authorization` header I could see this working simply.  I may have to do some research.

**Comment by princemaple** [comment]
Thanks for being rigorous Justin. Feel free to send a PR anyway, in case people want to help you with that. We can iterate on things.

**Comment by justindotpub** [comment]
If you don't mind humoring me, I'd like to think aloud to see if I'm on the right track.  :)

If I end up being correct that auth details are always ultimately provided in the `authorization` header, we could define a new behaviour (e.g. `Swoosh.AuthorizationHeaderProvider`) and then `Swoosh.Adapters.MsGraph` could accept something like `authorization_header_provider` config.  Then in `Swoosh.Adapters.MsGraph` I would use this module to call its `authorization_header` callback.  The advantage of this approach over an anonymous function would primarily be that we call a function on a module with a defined contract, and make it easier for others to reuse various implementations of this behaviour instead of always defining their own anonymous function.

Any corrections or additional thoughts to that?

**Comment by justindotpub** [comment]
Also note that I'm intentionally trying to keep all auth config params out of Swoosh, so there would not be any additional config options for, say, cases where someone is authenticating with username and password.

**Comment by justindotpub** [comment]
Good naming feedback, thank you.

I learned from https://learn.microsoft.com/en-us/graph/auth/auth-concepts that Graph auth always uses oauth 2.0 access tokens.  

Given that, I think I lean toward the behaviour just returning an oauth 2.0 access token.  The `MsGraph` code can still encode the token in the authorization header.

I have one more detail of my use case that I think impacts the definition of the behaviour.

The app I'm working on has multiple sets of email service configurations.  Each email service configuration corresponds to a brand that needs emails sent using the associated auth details.  When we request an oauth 2.0 token, we need the access token requester logic to know which email service we are requesting the token for, and ultimately that configuration comes from entries in a database.  What this means is that we need to pass in extra logic to the `request_access_token` callback to ensure that it has enough info to know if we already have a cached token for that configuration that hasn't expired.

Sorry if that is hard to follow.  Currently I have an access token cache GenServer per service configuration registered and looked up using the Registry and `:via` tuples, and all of that lookup logic is handled via that tiny little anonymous function `access_token_fn`.

All of this to say, if I went with the behaviour approach you're suggesting, I'm not sure how I would pass in that extra context about which service configuration the request was for so that it could be cached / looked up as appropriate.  The anonymous function lets me pass it in implicitly via the closure.

I feel like I'm back at wanting to stick with the anonymous fn.  🤔 

Thoughts?

**Comment by justindotpub** [comment]
Now that I think about it further, it's not just the caching and lookup logic that needs the extra parameter.  It is also the email configuration itself, like the other values I need in order to hit Microsoft's oauth 2.0 token requesting API.

The key point here is that, since I'm doing this dynamically based on values in a database, I can't just capture all of that logic in a module per email service configuration along with associated config.  A module can't capture it because I need to configure dynamically at runtime.

I think I'll just submit my PR and see what you think.  :)

**Comment by princemaple** [comment]
Hi Justin,

Thanks for the PR. It looks clean and simple. I'm more than happy to accept it in the current shape, with the code cleaned up. I think the auth part can totally be handled by a separate package which Swoosh doesn't have to care about. So this (either lambda or `{module, func, args}` tuple) is good enough for me.

@hermanius @geeksilva97 if you still use MS graph, could you give this a go? 

Cheers.

**Comment by princemaple** [comment]
Actually if we agree on complete leaving out the auth logic, we might as well just pass the token in as a string, dynamically, not through config. (lambda and MFA also work, too, via config or dynamically)

**Comment by justindotpub** [comment]
@princemaple I'm not sure I follow your last comment.  Are you saying we should pass it into `deliver`?  I assume not, since our function arity would no longer match the behaviour.  Or am I missing something?

**Comment by justindotpub** [comment]
Ah I see.  We have mailer config and dynamic config.  My vote is to move `access_token_fn` from mailer config to dynamic config.

MFA won't work for my specific use case, as mentioned above.

With dynamic config, either lambda or string works for me.

Given the info at https://erlef.github.io/security-wg/secure_coding_and_deployment_hardening/sensitive_data I lean toward sticking with a lambda though.

**Comment by justindotpub** [comment]
@princemaple if I understand you correctly that you are okay with the `access_token_fn` approach, I believe my PR is ready to go now.  I fixed the issue I was having with the tests.  Please let me know if you need anything else.  🙏

**Comment by princemaple** [comment]
> Ah I see. We have mailer config and dynamic config. My vote is to move `access_token_fn` from mailer config to dynamic config.

Yeah we merge all configs into one, so in `deliver` function, you can see everything from config file configs, module `use` configs, and configs in second argument of `deliver`.

> MFA won't work for my specific use case, as mentioned above.

This can be a simple `apply(m, f, a)`.

So something like this:

```elixir
auth_token =
  case config[:auth] do
    func when is_function(func, 0) ->
      func.()

    {m, f, a} when is_atom(m) and is_atom(f) and is_list(a) ->
      apply(m, f, a)

    token when is_binary(token) ->
      token

    nil ->
      raise "Missing required config :auth, it can be a function, a {mod, func, args} tuple, or a string"
  end
```

What do you think?

> I lean toward sticking with a lambda though.

Sure, a user can choose to pass a lambda.

**Comment by justindotpub** [comment]
Ah, that is a great idea.  Sorry, when you said MFA I thought you meant module, function, arity, not module, function, args.  Now I see the light.  😁 

I'll update my branch in the morning and let you know.  Thank you for your patience.

**Comment by justindotpub** [comment]
I decided to wrap up the changes before my bedtime.  :)  Take a look and let me know what you think.

**Comment by princemaple** [comment]
Thanks to @justindotpub , MsGraph will be released in the next version.
